spec/graphql/analysis/analyze_query_spec.rb

```diff
@@ -109,11 +115,103 @@ def call(memo, visit_type, irep_node)
       it "knows which fields are connections" do
         connection_counts = reduce_result.first
         expected_connection_counts = {
-          "field" => 5,
-          "connection" => 2
+          :field => 5,
+          :connection => 2
         }
         assert_equal expected_connection_counts, connection_counts
       end
     end
   end
+
+  describe ".visit_analyzers" do
+    class IdCatcher
+      def call(memo, visit_type, irep_node)
+        if visit_type == :enter
+          if irep_node.ast_node.name == "id"
+            raise GraphQL::AnalysisError.new("Don't use the id field.", irep_node.ast_node)
+          end
+        end
+        memo
+      end
+    end
+
+    class FlavorCatcher
+      def initial_value(query)
+        {
+          :errors => []
+        }
+      end
+
+      def call(memo, visit_type, irep_node)
+        if visit_type == :enter
+          if irep_node.ast_node.name == "flavor"
+            memo[:errors] << GraphQL::AnalysisError.new("Don't use the flavor field.", irep_node.ast_node)
+          end
+        end
+        memo
+      end
+
+      def final_value(memo)
+        memo[:errors]
+      end
+    end
+
+    let(:id_catcher) { IdCatcher.new }
+    let(:flavor_catcher) { FlavorCatcher.new }
+    let(:analyzers) { [id_catcher, flavor_catcher] }
+    let(:reduce_result) { GraphQL::Analysis.analyze_query(query, analyzers) }
+    let(:query) { GraphQL::Query.new(DummySchema, query_string) }
+    let(:query_string) {%|
+      {
+        cheese(id: 1) {
+          id
+          flavor
+        }
+      }
+    |}
+    let(:schema) { DummySchema }
+    let(:result) { schema.execute(query_string) }
+    let(:query_string) {%|
+      {
+        cheese(id: 1) {
+          id
+          flavor
+        }
+      }
+    |}
+
+    before do
+      @previous_query_analyzers = DummySchema.query_analyzers.dup
+      DummySchema.query_analyzers.clear
+      DummySchema.query_analyzers << id_catcher << flavor_catcher
+    end
+
+    after do
+      DummySchema.query_analyzers.clear
+      DummySchema.query_analyzers.push(*@previous_query_analyzers)
```

Comment:
No need to change, but I thought I'd share, since writing the previous test, I learned that these are equivalent: 

```ruby
# Modify array_1 by adding the members of array_2 to the end of it
array_1.push(*array_2)
array_1.concat(array_2)
```